content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx

```diff
@@ -0,0 +1,158 @@
+
+### Debugging `pthread`
+
+The [first error we ran into](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft#testing) was a memory corruption error with the `pthread` interface.
+After investigating the `pthread_join()` calls, I found that the threads were never even created correctly at all because `pthread_create()` uses the POSIX `mmap` system call, which never existed as we did not turn on virtual memory support (i.e., `ukvmem` alongside `posix-mmap`).
+> By the way, this is an error that would have never happened had I checked the return value of `pthread_create()` in the first place - lessons learned!
+
+### Debugging boot time virtual memory initialization
+
+There is a reason why we turned off virtual memory support when first porting the memory allocator.
+Unikraft's virtual memory interface requires the memory allocators to behave in certain ways that are not intrinsically supported by *mimalloc*.
+For example, this is what will happen when initializing the heap with virtual memory turned on:
+
+```c
+ /* In addition to paging, we have virtual address space management. We
+  * will thus also represent the heap as a dedicated VMA to enable
+  * on-demand paging for the heap. However, we have a chicken-egg
+  * problem here. This is because we need an allocator for allocating
+  * VMA objects but to do so, we need to create the heap VMA first.
+  * Theoretically, we could use a simple region allocator to bootstrap.
+  * But then, we would have to cope with VMA objects from different
+  * allocators. So instead, we just initialize a small heap using fixed
+  * mappings and then create the VMA on top. Afterwards, we add the
+  * remainder of the VMA to the allocator.
+  */
+```
+
+This is because Unikraft's Virtual Memory Areas (VMAs, described by `struct uk_vma`) are managed by a Virtual Address Space (VAS, described by `struct uk_vas`), which has one dedicated memory allocator assigned to it.
+So it would be impractical to allocate two VMAs under a same VAS using two different memory allocators.
+
+The code for this logic though, is customized for the buddy allocator:
+
+1. Initialize a physically-mapped small heap size of 16 pages
+1. Initialize the allocator based on this heap
+1. Initialize the VAS
+1. Map the rest of the heap as an anonymous VMA
+1. Call `uk_alloc_addmem()` to let the buddy allocator know that this anonymous region that is at its disposal
+
+This design accommodates the fact that the buddy allocator needs initial memory space to store its metadata and that it needs information about the total available memory for initializing its free lists.
+However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs are not yet ready.
```

Comment:
```suggestion
However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs is not yet ready.
```